lib/Sema/CSBindings.cpp

```diff
@@ -366,11 +366,49 @@ void PotentialBindings::inferTransitiveBindings(
   }
 }
 
+// If potential binding type variable or any of it's adjacents represents the
+// argument in an argument conversion constraint located directly on an
+// autoclosure parameter.
+static bool
+isClosureInAutoClosureArgumentConversion(PotentialBindings &bindings) {
+  auto &cs = bindings.CS;
+  auto isClosureVarInAutoClosureContext = [&](TypeVariableType *typeVar) {
+    if (!typeVar->getImpl().isClosureType())
+      return false;
+
+    if (auto *argLocator =
+            cs.getClosureArgumentConversionLocatorInfo(typeVar)) {
+      return argLocator->isLastElement<LocatorPathElt::AutoclosureResult>();
+    }
+    return false;
+  };
+
+  if (isClosureVarInAutoClosureContext(bindings.TypeVar))
+    return true;
+
+  return llvm::any_of(bindings.AdjacentVars, isClosureVarInAutoClosureContext);
+}
+
 void PotentialBindings::finalize(
     llvm::SmallDenseMap<TypeVariableType *, PotentialBindings>
         &inferredBindings) {
   inferTransitiveProtocolRequirements(inferredBindings);
   inferTransitiveBindings(inferredBindings);
+
+  // For autoclosure parameters if we have a closure argument which could
+  // default to `() -> $T`, we avoid infering defaultable binding because
+  // an autoclosure cannot accept a closure paramenter unless the result `$T`
+  // is bound to a function type via another contextual binding. Also consider
+  // adjacent vars because they can also default transitively.
+  if (isClosureInAutoClosureArgumentConversion(*this)) {
```

Comment:
Let's consider when this can happen:

- closure is used as an argument to call to a function or operator where parameter is of type `@autoclosure () -> T`
- result of an @autoclosure has to be a type variable, otherwise closure would be able to infer a type and avoid defaulting;

Binding inference already collects all subtype, supertype and equivalence relationships where other side is a type variable. I think you can leverage that information (you need `SubtypeOf` collection associated with this type variable) to make a decision about removal of default type.

Reply:
> Binding inference already collects all subtype, supertype and equivalence relationships where other side is a type variable.

Ah I didn't know that. Thank you! 
I tried something similar because I note there was relation between `$T3` and  `$T4 in ()-> $T4` where the later would point to an auto closure, but I was trying to get this information through the constraint graph... and couldn't make it work 🤦 

Thank you for the info! As always, I learned something interesting =]

I've made the changes, so let me know what you think